src/adapter/src/coord/sequencer/inner.rs

```diff
@@ -3312,15 +3342,179 @@ impl Coordinator {
+            self.least_valid_read(&id_bundle)
+        };
+        // let enable_consolidate_after_union_negate = self
+        //     .catalog()
+        //     .system_config()
+        //     .enable_consolidate_after_union_negate();
+
+        // Create a transient catalog item
+        // -------------------------------
+
+        let mut transient_items = BTreeMap::new();
+        transient_items.insert(exported_sink_id, {
+            TransientItem::new(
+                Some(full_name.to_string()),
+                Some(full_name.item.to_string()),
+                Some(column_names.iter().map(|c| c.to_string()).collect()),
+            )
+        });
+
+        // Execute the various stages of the optimization pipeline
+        // -------------------------------------------------------
+
+        // Trace the pipeline input under `optimize/raw`.
+        tracing::span!(target: "optimizer", Level::TRACE, "raw").in_scope(|| {
+            trace_plan(&raw_plan);
+        });
+
+        // Execute the `optimize/hir_to_mir` stage.
+        let decorrelated_plan = catch_unwind(broken, "hir_to_mir", || {
+            raw_plan.optimize_and_lower(&OptimizerConfig {})
+        })?;
+
+        // Execute the `optimize/local` stage.
+        let optimized_plan = catch_unwind(broken, "local", || {
+            tracing::span!(target: "optimizer", Level::TRACE, "local").in_scope(|| {
+                let optimized_plan = self.view_optimizer.optimize(decorrelated_plan);
+                if let Ok(ref optimized_plan) = optimized_plan {
+                    trace_plan(optimized_plan.as_inner());
+                }
+                optimized_plan.map_err(AdapterError::from)
+            })
+        })?;
+
+        // Execute the `optimize/global` stage.
+        let (mut df, df_metainfo) = catch_unwind(broken, "global", || {
+            let mut dataflow_builder =
+                DataflowBuilder::new(self.catalog().state(), compute_instance);
+            dataflow_builder.build_materialized_view(
+                exported_sink_id,
+                internal_view_id,
+                debug_name,
+                &optimized_plan,
+                &RelationDesc::new(optimized_plan.typ(), column_names),
+            )
+        })?;
+
+        df.set_as_of(as_of);
+
+        // If the only outputs of the dataflow are sinks, we might be able to
+        // turn off the computation early, if they all have non-trivial
+        // `up_to`s.
+        //
+        // TODO: This should always be the case here so we can demote
+        // the outer index to a soft assert.
+        if df.index_exports.is_empty() {
+            df.until = Antichain::from_elem(Timestamp::MIN);
+            for (_, sink) in &df.sink_exports {
+                df.until.join_assign(&sink.up_to);
+            }
+        }
+
+        // Collect the list of indexes used by the dataflow at this point
+        let used_indexes = UsedIndexes::new(
```

Comment:
This could be factored out into a function, as there are 3 copies now.

Reply:
I wanted to revisit this separately. I don't think we should be calling `prune_and_annotate_dataflow_index_imports` as part of the `optimize_dataflow` call.

Rather, given a `DataflowDescription` we should be able to do this ad-hoc just before rendering the corresponding plan. At the moment we do this once towards the end of the optimizer pipeline but carry this information for plans that occur before and after that.

Reply:
> produces possibly incorrect EXPLAIN output (an index might be printed even though the associated plan structure does not reveal that we would actually use it).

So, if I understand correctly, this is the "freezing" problem that we discussed some time ago in a Friday meeting. (Note that `optimize_dataflow_monotonic` also has the same problem that if some transformation happens to the plan afterwards, then it might possibly go out of sync with the plan.) But I'm not sure if pulling these out of `optimize_dataflow` is the best solution, because then these functions (`prune_and_annotate_dataflow_index_imports` and `optimize_dataflow_monotonic`) would need to be called from several different places, which would also complicate the code, and would introduce the danger that some code path that calls `optimize_dataflow` forgets to call these. Or, if there will be a new wrapper method of `optimize_dataflow` that calls these functions (maybe as part of the new optimizer interface), then the problem is just shifted to that wrapper method being forbidden to do transforms afterwards. If we want a very clean solution, then we can create a `FrozenMirRelationExpr` (as you mentioned before), where the structure is not allowed to change anymore, just these metainfos like monotonic flag or used indexes. But anyhow, just putting a big warning comment in `optimize_dataflow` for now should almost certainly prevent us from making the mistake of putting some transform after these function calls.

Reply:
In general `prune_and_annotate_dataflow_index_imports` does two things:

1. Annotates index imports with their usage type.
2. Prunes index types that are not used.

If we can split the two functions, we can just call (2) with the output of (1) in dataflow rendering but then drop the annotations.

At the moment `DataflowMetainfo` contains both:

1. fields (`notices`) that represent the cumulative effect of all optimizer stages traversed by the `optimize_dataflow` call, and 2. fields (`index_usage_types`) that represent a specific optimizer stage and are just memoized function results that can be always re-comupted from the `DataflowDescription`.

It will be easier to deal with `DataflowMetainfo` instances if we commit that they serve only one of these two purposes and find a different solution for the other one.

Reply:
> But anyhow, just putting a big warning comment in `optimize_dataflow` for now should almost certainly prevent us from making the mistake of putting some transform after these function calls.

I was thinking more about anomalies in the opposite order. Assume that you have a set of available indexes `I` and from that you determine a subset of used indexes (`U`) in optimization stage `j`. If for some reason in some stage `i < j` your plan was using indexes that are in `I - U`, the explain output will not render this correctly.

This is a niche problem though, and it could be that the set of used indexes can only grow monotonically as we advance through the optimization stages, so this situation cannot happen in practice.

Reply:
> If for some reason in some stage i < j your plan was using indexes that are in I - U, the explain output will not render this correctly.

Hmm, yes, we'll need to fix this at some point.